File: Hash_Table/problem127/problem127.py
from collections import deque
from typing import List


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        words = set(wordList)
        if endWord not in words:
            return 0

        q = deque([(beginWord, 1)])
        letters = "abcdefghijklmnopqrstuvwxyz"

        while q:
            w, d = q.popleft()
            if w == endWord:
                return d
            for i in range(len(w)):
                for c in letters:
                    if c == w[i]:
                        continue
                    nw = w[:i] + c + w[i + 1 :]
                    if nw in words:
                        words.remove(nw)
                        q.append((nw, d + 1))

        return 0

File: Hash_Table/problem127/problem127.cpp
#include <deque>
#include <string>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> words(wordList.begin(), wordList.end());
        if (!words.count(endWord)) {
            return 0;
        }

        deque<pair<string, int>> q;
        q.emplace_back(beginWord, 1);

        while (!q.empty()) {
            auto cur = q.front().first;
            const int d = q.front().second;
            q.pop_front();
            if (cur == endWord) {
                return d;
            }
            for (int i = 0; i < static_cast<int>(cur.size()); ++i) {
                for (char c = 'a'; c <= 'z'; ++c) {
                    if (c == cur[i]) {
                        continue;
                    }
                    string nw = cur;
                    nw[i] = c;
                    if (words.count(nw)) {
                        words.erase(nw);
                        q.emplace_back(std::move(nw), d + 1);
                    }
                }
            }
        }
        return 0;
    }
};
